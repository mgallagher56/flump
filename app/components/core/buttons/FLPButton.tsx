import React from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import * as pressable from '@zag-js/pressable';
import type { PressEvent } from '@zag-js/pressable/dist/pressable.types';
import { normalizeProps, useMachine } from '@zag-js/react';

interface FLPButtonProps extends ButtonProps {
  preventFocusOnPress?: boolean;
  onPress?: (e: PressEvent) => void;
  onLongPress?: (e: PressEvent) => void;
}

const FLPButton: React.FC<React.PropsWithChildren<FLPButtonProps>> = ({
  children,
  colorScheme = 'blue',
  isDisabled,
  preventFocusOnPress,
  variant = 'solid',
  onLongPress,
  onPress,
  isLoading
}) => {
  const [state, send] = useMachine(
    pressable.machine({
      id: 'pressableBaseButton',
      onPress,
      onLongPress,
      disabled: isDisabled,
      preventFocusOnPress
    })
  );

  const api = pressable.connect(state, send, normalizeProps);

  return (
    <Button {...api.pressableProps} variant={variant} colorScheme={colorScheme} isLoading={isLoading}>
      {children}
    </Button>
  );
};

export default FLPButton;
