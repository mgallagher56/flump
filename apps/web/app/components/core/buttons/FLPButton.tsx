import { type FC, type PropsWithChildren } from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';

interface FLPButtonProps extends ButtonProps {
  padding?: number;
}

const FLPButton: FC<PropsWithChildren<FLPButtonProps>> = ({
  colorScheme = 'blue',
  disabled,
  isLoading,
  padding,
  variant = 'solid',
  onClick,
  ...buttonProps
}) => {
  return (
    <Button
      colorScheme={colorScheme}
      isDisabled={disabled}
      isLoading={isLoading}
      padding={padding}
      variant={variant}
      onClick={onClick}
      width={'max-content'}
      {...buttonProps}
    >
      {buttonProps.children}
    </Button>
  );
};

export default FLPButton;
