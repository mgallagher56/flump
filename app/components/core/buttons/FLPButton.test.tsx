import { fireEvent, render, screen } from '@testing-library/react';
import { vi } from 'vitest';

import FLPButton from './FLPButton';

describe('FLPButton', () => {
  it('should render correctly', () => {
    const { container } = render(<FLPButton>FLPButton</FLPButton>);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with props', () => {
    const { container } = render(
      <FLPButton colorScheme="green" isDisabled preventFocusOnPress variant="outline" isLoading>
        FLPButton
      </FLPButton>
    );
    expect(container).toMatchSnapshot();
  });

  it('should call onPress when clicked', () => {
    const onPress = vi.fn();
    render(<FLPButton onPress={onPress}>FLPButton</FLPButton>);
    fireEvent.click(screen.getByText('FLPButton'));

    expect(onPress).toHaveBeenCalled();
  });

  it('should not call onPress when clicked and isDisabled is true', () => {
    const onPress = vi.fn();
    render(
      <FLPButton onPress={onPress} isDisabled>
        FLPButton
      </FLPButton>
    );
    fireEvent.click(screen.getByText('FLPButton'));
    expect(onPress).not.toHaveBeenCalled();
  });

  it('should not call onPress when clicked and preventFocusOnPress is true', () => {
    const onPress = vi.fn();
    render(
      <FLPButton onPress={onPress} preventFocusOnPress>
        FLPButton
      </FLPButton>
    );
    fireEvent.click(screen.getByText('FLPButton'));
    expect(onPress).toHaveBeenCalled();
  });

  it('should not call onPress when clicked and isLoading is true', () => {
    const onPress = vi.fn();
    render(
      <FLPButton onPress={onPress} isLoading>
        FLPButton
      </FLPButton>
    );
    fireEvent.click(screen.getByText('FLPButton'));
    expect(onPress).not.toHaveBeenCalled();
  });
});
