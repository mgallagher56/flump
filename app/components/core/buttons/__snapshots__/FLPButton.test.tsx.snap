// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FLPButton > should not call onClick when clicked and isDisabled is true 1`] = `
<body>
  <div>
    <button
      class="chakra-button css-4xx2wk"
      data-disabled=""
      disabled=""
      type="button"
    >
      FLPButton
    </button>
  </div>
</body>
`;

exports[`FLPButton > should render correctly 1`] = `
<div>
  <button
    class="chakra-button css-4xx2wk"
    type="button"
  >
    FLPButton
  </button>
</div>
`;

exports[`FLPButton > should render correctly with props 1`] = `
<div>
  <button
    class="chakra-button css-4xx2wk"
    data-disabled=""
    data-loading=""
    disabled=""
    type="button"
  >
    <div
      class="chakra-button__spinner chakra-button__spinner--start css-1vsp78f"
    >
      <div
        class="chakra-spinner css-1plzyg6"
      >
        <span
          class="css-8b45rq"
        >
          Loading...
        </span>
      </div>
    </div>
    <span
      class="css-q7lffx"
    >
      FLPButton
    </span>
  </button>
</div>
`;
